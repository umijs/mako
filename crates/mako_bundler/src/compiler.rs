use crate::{
    build::build::BuildParam, config::Config, context::Context, generate::generate::GenerateParam,
};

pub struct Compiler {
    pub context: Context,
}

impl Compiler {
    pub fn new(config: Config) -> Self {
        let context = Context::new(config);
        Self { context }
    }

    pub fn run(&mut self) {
        self.build(&BuildParam { files: None });
        self.generate(&GenerateParam { write: true });
    }

    pub fn _watch(&self) {}
}
