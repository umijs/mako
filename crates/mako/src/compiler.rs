use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex, RwLock};
use std::time::Instant;

use anyhow::{anyhow, Error, Result};
use colored::Colorize;
use regex::Regex;
use swc_core::common::sync::Lrc;
use swc_core::common::{Globals, SourceMap, DUMMY_SP};
use swc_core::ecma::ast::Ident;
use swc_node_comments::SwcComments;
use tracing::debug;

use crate::ast::comments::Comments;
use crate::ast::file::win_path;
use crate::config::{Config, Mode, ModuleIdStrategy, OutputMode};
use crate::generate::chunk_graph::ChunkGraph;
use crate::generate::optimize_chunk::OptimizeChunksInfo;
use crate::module_graph::ModuleGraph;
use crate::plugin::{Plugin, PluginDriver, PluginGenerateEndParams};
use crate::plugins;
use crate::resolve::{get_resolvers, Resolvers};
use crate::share::helpers::SWC_HELPERS;
use crate::stats::StatsInfo;
use crate::utils::id_helper::{assign_numeric_ids, compare_modules_by_incoming_edges};
use crate::utils::{thread_pool, ParseRegex};

pub struct Context {
    pub module_graph: RwLock<ModuleGraph>,
    pub chunk_graph: RwLock<ChunkGraph>,
    pub assets_info: Mutex<HashMap<String, String>>,
    pub modules_with_missing_deps: RwLock<Vec<String>>,
    pub config: Config,
    pub numeric_ids_map: RwLock<HashMap<String, usize>>,
    pub args: Args,
    pub root: PathBuf,
    pub meta: Meta,
    pub plugin_driver: PluginDriver,
    pub stats_info: StatsInfo,
    pub resolvers: Resolvers,
    pub static_cache: RwLock<MemoryChunkFileCache>,
    pub optimize_infos: Mutex<Option<Vec<OptimizeChunksInfo>>>,
}

#[derive(Default)]
pub struct MemoryChunkFileCache {
    content_map: HashMap<String, (Vec<u8>, u64)>,
    root: Option<PathBuf>,
}

impl MemoryChunkFileCache {
    pub fn new(root: Option<PathBuf>) -> Self {
        Self {
            content_map: HashMap::new(),
            root,
        }
    }

    pub fn write<T: AsRef<str>>(&mut self, path: T, content: Vec<u8>, hash: u64) -> Result<()> {
        let str = path.as_ref();

        if let Some((_, in_mem_hash)) = self.content_map.get(str) {
            if *in_mem_hash != hash {
                self.write_to_disk(str, &content)?;
            }
        } else {
            self.write_to_disk(str, &content)?;
        }
        self.content_map
            .insert(path.as_ref().to_string(), (content, hash));
        Ok(())
    }

    pub fn read<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        self.content_map
            .get(path.as_ref())
            .map(|(content, _)| content.clone())
    }

    fn write_to_disk<T: AsRef<str>>(&self, path: T, content: &[u8]) -> Result<()> {
        if let Some(root) = &self.root {
            let path = root.join(path.as_ref());
            fs::write(path, content)?;
        }
        Ok(())
    }
}

#[derive(Default)]
pub struct Args {
    pub watch: bool,
}

impl Context {
    pub fn write_static_content<T: AsRef<str>>(
        &self,
        path: T,
        content: Vec<u8>,
        hash: u64,
    ) -> Result<()> {
        let mut map = self.static_cache.write().unwrap();
        map.write(path, content, hash)
    }

    pub fn get_static_content<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        let map = self.static_cache.read().unwrap();
        map.read(path)
    }
}

impl Default for Context {
    fn default() -> Self {
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        let config: Config = Default::default();
        let resolvers = get_resolvers(&config);
        Self {
            config,
            args: Args { watch: false },
            root: PathBuf::from(""),
            module_graph: RwLock::new(ModuleGraph::new()),
            chunk_graph: RwLock::new(ChunkGraph::new()),
            assets_info: Mutex::new(HashMap::new()),
            modules_with_missing_deps: RwLock::new(Vec::new()),
            meta: Meta::new(),
            plugin_driver: Default::default(),
            stats_info: StatsInfo::new(),
            resolvers,
            optimize_infos: Mutex::new(None),
            static_cache: Default::default(),
            numeric_ids_map: RwLock::new(numeric_ids_map),
        }
    }
}

pub struct Meta {
    pub script: ScriptMeta,
    pub css: CssMeta,
}

impl Meta {
    pub fn new() -> Self {
        Self {
            script: ScriptMeta::new(),
            css: CssMeta::new(),
        }
    }
}

impl Default for Meta {
    fn default() -> Self {
        Self::new()
    }
}

pub struct ScriptMeta {
    pub cm: Lrc<SourceMap>,
    pub origin_comments: RwLock<Comments>,
    pub output_comments: RwLock<Comments>,
    pub globals: Globals,
    // These idents may be used in other places, such as transform_async_module
    pub module_ident: Ident,
    pub exports_ident: Ident,
    pub require_ident: Ident,
}

impl ScriptMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            origin_comments: Default::default(),
            output_comments: Default::default(),
            globals: Globals::default(),
            module_ident: build_ident("module"),
            exports_ident: build_ident("exports"),
            require_ident: build_ident("__mako_require__"),
        }
    }
}

fn build_ident(ident: &str) -> Ident {
    Ident {
        ctxt: Default::default(),
        span: DUMMY_SP,
        sym: ident.into(),
        optional: false,
    }
}

pub struct CssMeta {
    pub cm: Lrc<SourceMap>,
    pub globals: Globals,
    pub comments: SwcComments,
}

impl CssMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            globals: Globals::default(),
            comments: Default::default(),
        }
    }
}

impl Context {
    pub fn emit_assets(&self, origin_path: String, output_path: String) {
        let mut assets_info = self.assets_info.lock().unwrap();
        assets_info.insert(origin_path, output_path);
    }
}

pub struct Compiler {
    pub context: Arc<Context>,
}

impl Compiler {
    pub fn new(
        config: Config,
        root: PathBuf,
        args: Args,
        extra_plugins: Option<Vec<Arc<dyn Plugin>>>,
    ) -> Result<Self> {
        if !root.is_absolute() {
            return Err(anyhow!("root path must be absolute"));
        }

        let root = PathBuf::from(win_path(root.to_str().unwrap()));

        // why add plugins before builtin plugins?
        // because plugins like less-loader need to be added before assets plugin
        // TODO: support plugin orders
        let mut plugins: Vec<Arc<dyn Plugin>> = vec![];
        if let Some(extra_plugins) = extra_plugins {
            plugins.extend(extra_plugins);
        }
        let builtin_plugins: Vec<Arc<dyn Plugin>> = vec![
            // features
            Arc::new(plugins::manifest::ManifestPlugin {}),
            Arc::new(plugins::copy::CopyPlugin {}),
            Arc::new(plugins::import::ImportPlugin {}),
            // file types
            Arc::new(plugins::context_module::ContextModulePlugin {}),
            Arc::new(plugins::runtime::MakoRuntime {}),
            Arc::new(plugins::invalid_webpack_syntax::InvalidWebpackSyntaxPlugin {}),
            Arc::new(plugins::hmr_runtime::HMRRuntimePlugin {}),
            Arc::new(plugins::wasm_runtime::WasmRuntimePlugin {}),
            Arc::new(plugins::async_runtime::AsyncRuntimePlugin {}),
            Arc::new(plugins::emotion::EmotionPlugin {}),
            Arc::new(plugins::tree_shaking::FarmTreeShake {}),
            Arc::new(plugins::detect_circular_dependence::LoopDetector {}),
        ];
        plugins.extend(builtin_plugins);

        let mut config = config;

        if config.mode == Mode::Production && config.experimental.imports_checker {
            plugins.push(Arc::new(plugins::imports_checker::ImportsChecker {}));
        }

        if let Some(progress) = &config.progress {
            plugins.push(Arc::new(plugins::progress::ProgressPlugin::new(
                plugins::progress::ProgressPluginOptions {
                    prefix: "Mako".to_string(),
                    template:
                        "‚óè {prefix:.bold} {bar:25.green/white.dim} ({percent}%) {wide_msg:.dim}"
                            .to_string(),
                    progress_chars: progress.progress_chars.clone(),
                },
            )));
        }

        if let Some(duplicate_package_checker) = &config.check_duplicate_package {
            plugins.push(Arc::new(
                plugins::duplicate_package_checker::DuplicatePackageCheckerPlugin::new()
                    .show_help(duplicate_package_checker.show_help)
                    .emit_error(duplicate_package_checker.emit_error)
                    .verbose(duplicate_package_checker.verbose),
            ));
        }

        if config.experimental.require_context {
            plugins.push(Arc::new(plugins::require_context::RequireContextPlugin {}))
        }

        if config.output.mode == OutputMode::Bundless {
            plugins.insert(
                0,
                Arc::new(plugins::bundless_compiler::BundlessCompilerPlugin {}),
            );
        }

        if std::env::var("DEBUG_GRAPH").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::graphviz::Graphviz {}));
        }

        if args.watch && std::env::var("SSU").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::ssu::SUPlus::new()));
        }

        if args.watch && config.experimental.central_ensure {
            plugins.push(Arc::new(plugins::central_ensure::CentralChunkEnsure {}));
        }

        if let Some(minifish_config) = &config._minifish {
            let inject = if let Some(inject) = &minifish_config.inject {
                let mut map = HashMap::new();

                for (k, ii) in inject.iter() {
                    map.insert(
                        k.clone(),
                        plugins::minifish::Inject {
                            from: ii.from.clone(),
                            name: k.clone(),
                            named: ii.named.clone(),
                            namespace: ii.namespace,
                            exclude: ii.exclude.parse_into_regex()?,
                            include: ii.include.parse_into_regex()?,
                            prefer_require: ii.prefer_require.map_or(false, |v| v),
                        },
                    );
                }
                Some(map)
            } else {
                None
            };

            plugins.insert(
                0,
                Arc::new(plugins::minifish::MinifishPlugin {
                    mapping: minifish_config.mapping.clone(),
                    meta_path: minifish_config.meta_path.clone(),
                    inject,
                }),
            );
        }

        let ignores = config
            .ignores
            .iter()
            .map(|ignore| Regex::new(ignore).map_err(Error::new))
            .collect::<Result<Vec<Regex>>>()?;
        plugins.push(Arc::new(plugins::ignore::IgnorePlugin { ignores }));

        let plugin_driver = PluginDriver::new(plugins);

        plugin_driver.modify_config(&mut config, &root, &args)?;

        let resolvers = get_resolvers(&config);
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        Ok(Self {
            context: Arc::new(Context {
                static_cache: if config.write_to_disk {
                    RwLock::new(MemoryChunkFileCache::new(Some(config.output.path.clone())))
                } else {
                    Default::default()
                },
                config,
                args,
                root,
                module_graph: RwLock::new(ModuleGraph::new()),
                chunk_graph: RwLock::new(ChunkGraph::new()),
                assets_info: Mutex::new(HashMap::new()),
                modules_with_missing_deps: RwLock::new(Vec::new()),
                meta: Meta::new(),
                plugin_driver,
                numeric_ids_map: RwLock::new(numeric_ids_map),
                stats_info: StatsInfo::new(),
                resolvers,
                optimize_infos: Mutex::new(None),
            }),
        })
    }

    pub fn compile(&self) -> Result<()> {
        // ÂÖàÊ∏ÖÁ©∫ dist ÁõÆÂΩï
        if self.context.config.clean {
            self.clean_dist()?;
        }

        let t_compiler = Instant::now();
        let start_time = chrono::Local::now().timestamp_millis();
        let building_with_message = format!(
            "Building with {} for {}...",
            "mako".to_string().cyan(),
            self.context.config.mode
        )
        .green();
        println!("{}", building_with_message);
        {
            crate::mako_profile_scope!("Build Stage");
            let files = self
                .context
                .config
                .entry
                .values()
                .map(|entry| {
                    let mut entry = entry.to_string_lossy().to_string();
                    let is_browser = matches!(
                        self.context.config.platform,
                        crate::config::Platform::Browser
                    );
                    let watch = self.context.args.watch;
                    let hmr = self.context.config.hmr.is_some();
                    if is_browser && watch && hmr {
                        entry = format!("{}?hmr", entry);
                    }
                    crate::ast::file::File::new_entry(entry, self.context.clone())
                })
                .collect();
            self.context.plugin_driver.build_start(&self.context)?;

            self.build(files)?;

            debug!("start after build");

            self.context
                .plugin_driver
                .after_build(&self.context, self)?;
        }

        self.context.plugin_driver.before_generate(&self.context)?;

        if let ModuleIdStrategy::Numeric = self.context.config.module_id_strategy {
            let module_graph = self.context.module_graph.read().unwrap();
            assign_numeric_ids(
                module_graph.modules(),
                |a, b| compare_modules_by_incoming_edges(&module_graph, &a.id, &b.id),
                |module, id| {
                    let mut numeric_ids_map = self.context.numeric_ids_map.write().unwrap();
                    // reserved ten indexes for swc helper and others runtime module
                    numeric_ids_map.insert(module.id.id.clone(), id + 10);
                },
            )
        }

        let result = {
            crate::mako_profile_scope!("Generate Stage");
            // need to put all rayon parallel iterators run in the existed scope, or else rayon
            // will create a new thread pool for those parallel iterators
            thread_pool::scope(|_| self.generate())
        };
        let t_compiler_duration = t_compiler.elapsed();
        match result {
            Ok(mut stats) => {
                stats.start_time = start_time;
                stats.end_time = chrono::Local::now().timestamp_millis();
                println!(
                    "{}",
                    format!(
                        "‚úì Built in {}",
                        format!("{}ms", t_compiler_duration.as_millis()).bold()
                    )
                    .green()
                );
                if !self.context.args.watch {
                    println!("{}", "Complete!".bold());
                }
                let params = PluginGenerateEndParams {
                    is_first_compile: true,
                    time: t_compiler.elapsed().as_millis() as i64,
                    stats,
                };
                self.context
                    .plugin_driver
                    .generate_end(&params, &self.context)?;
                self.context.plugin_driver.write_bundle(&self.context)?;
                Ok(())
            }
            Err(e) => Err(e),
        }
    }

    pub fn full_hash(&self) -> u64 {
        crate::mako_profile_function!();
        let cg = self.context.chunk_graph.read().unwrap();
        let mg = self.context.module_graph.read().unwrap();
        cg.full_hash(&mg)
    }

    fn clean_dist(&self) -> Result<()> {
        // compiler ÂâçÊ∏ÖÈô§ distÔºåÂ¶ÇÊûúÂêéÁª≠ dev ÁéØÂ¢É‰∏çÂú® output_path ÈáåÔºåÈúÄË¶ÅÂÜçË°•‰∏ä dev ÁöÑÈÄªËæë
        let output_path = &self.context.config.output.path;
        if fs::metadata(output_path).is_ok() {
            fs::remove_dir_all(output_path)?;
        }
        Ok(())
    }
}
