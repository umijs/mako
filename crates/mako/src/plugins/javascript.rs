use std::sync::Arc;

use mako_core::anyhow::Result;
use mako_core::swc_atoms::js_word;
use mako_core::swc_common::collections::AHashSet;
use mako_core::swc_common::sync::Lrc;
use mako_core::swc_common::{self, Mark, GLOBALS};
use mako_core::swc_ecma_ast::{
    CallExpr, Callee, Expr, Id, Ident, Import, Lit, MemberExpr, MemberProp, MetaPropExpr,
    MetaPropKind, Module, ModuleDecl, NewExpr, Str,
};
use mako_core::swc_ecma_utils::collect_decls;
use mako_core::swc_ecma_visit::{Visit, VisitWith};

use super::css::is_url_ignored;
use crate::ast::build_js_ast;
use crate::compiler::Context;
use crate::load::{read_content, Content};
use crate::module::{Dependency, ModuleAst, ResolveType};
use crate::plugin::{Plugin, PluginDepAnalyzeParam, PluginLoadParam, PluginParseParam};

pub struct JavaScriptPlugin {}

impl Plugin for JavaScriptPlugin {
    fn name(&self) -> &str {
        "javascript"
    }

    fn load(&self, param: &PluginLoadParam, context: &Arc<Context>) -> Result<Option<Content>> {
        if param
            .task
            .is_match(vec!["js", "jsx", "ts", "tsx", "cjs", "mjs"])
        {
            if param.task.is_entry && param.task.request.has_query("hmr") {
                let port = &context.config.hmr_port.to_string();
                let host = &context.config.hmr_host.to_string();
                let host = if host == "0.0.0.0" { "127.0.0.1" } else { host };
                let content = format!(
                    "module.exports = require(\"{}\");",
                    param.task.request.path.as_str()
                );
                let content = format!(
                    "{}\n{}\n",
                    include_str!("../runtime/runtime_hmr_entry.js"),
                    content,
                )
                .replace("__PORT__", port)
                .replace("__HOST__", host);
                return Ok(Some(Content::Js(content)));
            }

            let content = read_content(param.task.request.path.as_str())?;
            return Ok(Some(Content::Js(content)));
        }
        Ok(None)
    }

    fn parse(&self, param: &PluginParseParam, context: &Arc<Context>) -> Result<Option<ModuleAst>> {
        if let Content::Js(content) = param.content {
            let ast = build_js_ast(&param.task.request.path, content, context)?;
            return Ok(Some(ModuleAst::Script(ast)));
        }
        Ok(None)
    }

    fn analyze_deps(
        &self,
        param: &mut PluginDepAnalyzeParam,
        context: &Arc<Context>,
    ) -> Result<Option<Vec<Dependency>>> {
        if let ModuleAst::Script(script) = param.ast {
            let mut visitor = DepCollectVisitor::new(script.unresolved_mark);
            GLOBALS.set(&context.meta.script.globals, || {
                script.ast.visit_with(&mut visitor);
                Ok(Some(visitor.dependencies))
            })
        } else {
            Ok(None)
        }
    }
}

struct DepCollectVisitor {
    bindings: Lrc<AHashSet<Id>>,
    dependencies: Vec<Dependency>,
    order: usize,
    unresolved_mark: Mark,
}

impl DepCollectVisitor {
    fn new(unresolved_mark: Mark) -> Self {
        Self {
            bindings: Default::default(),
            dependencies: vec![],
            // start with 1
            // 0 for swc helpers
            order: 1,
            unresolved_mark,
        }
    }
    fn bind_dependency(
        &mut self,
        source: String,
        resolve_type: ResolveType,
        span: Option<swc_common::Span>,
    ) {
        self.dependencies.push(Dependency {
            source,
            resolve_as: None,
            order: self.order,
            resolve_type,
            span,
        });
        self.order += 1;
    }
}

impl Visit for DepCollectVisitor {
    fn visit_module(&mut self, module: &Module) {
        self.bindings = Lrc::new(collect_decls(module));
        module.visit_children_with(self);
    }
    fn visit_module_decl(&mut self, n: &ModuleDecl) {
        match n {
            ModuleDecl::Import(import) => {
                if import.type_only {
                    return;
                }
                let src = import.src.value.to_string();
                self.bind_dependency(src, ResolveType::Import, Some(import.src.span));
            }
            ModuleDecl::ExportNamed(export) => {
                if let Some(src) = &export.src {
                    self.bind_dependency(
                        src.value.to_string(),
                        ResolveType::ExportNamed,
                        Some(src.span),
                    );
                }
            }
            ModuleDecl::ExportAll(export) => {
                let src = export.src.value.to_string();
                self.bind_dependency(src, ResolveType::ExportAll, Some(export.src.span));
            }
            _ => {}
        }
        // export function xxx() {} 里可能包含 require 或 import()
        n.visit_children_with(self);
    }
    fn visit_call_expr(&mut self, expr: &CallExpr) {
        if is_commonjs_require(expr, &self.unresolved_mark) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::Require, Some(expr.span));
                return;
            }
        } else if is_dynamic_import(expr) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::DynamicImport, Some(expr.span));
                return;
            }
        }
        expr.visit_children_with(self);
    }

    // Web workers
    fn visit_new_expr(&mut self, new_expr: &NewExpr) {
        if let Some(str) = resolve_web_worker(new_expr, self.unresolved_mark) {
            self.bind_dependency(str.value.to_string(), ResolveType::Worker, None);
        }

        new_expr.visit_children_with(self);
    }
}

pub fn resolve_web_worker(new_expr: &NewExpr, unresolved_mark: Mark) -> Option<&Str> {
    if !new_expr.args.as_ref().is_some_and(|args| !args.is_empty()) || !new_expr.callee.is_ident() {
        return None;
    }

    if let box Expr::Ident(Ident { span, sym, .. }) = &new_expr.callee {
        // `Worker` must be unresolved
        if sym == "Worker" && (span.ctxt.outer() == unresolved_mark) {
            let args = new_expr.args.as_ref().unwrap();

            // new Worker(new URL(''), base);
            if let Expr::New(new_expr) = &*args[0].expr {
                if !new_expr.args.as_ref().is_some_and(|args| !args.is_empty())
                    || !new_expr.callee.is_ident()
                {
                    return None;
                }

                if let box Expr::Ident(Ident { span, sym, .. }) = &new_expr.callee {
                    if sym == "URL" && (span.ctxt.outer() == unresolved_mark) {
                        // new URL(''); 仅第一个参数为字符串字面量, 第二个参数为 import.meta.url 时添加依赖
                        let args = new_expr.args.as_ref().unwrap();

                        if args.get(1).is_none() || !is_import_meta_url(&args.get(1).unwrap().expr)
                        {
                            return None;
                        }

                        if let box Expr::Lit(Lit::Str(ref str)) = &args[0].expr {
                            if !is_url_ignored(&str.value) {
                                return Some(str);
                            }
                        }
                    }
                }
            }
        }
    }

    None
}

pub fn is_import_meta_url(expr: &Expr) -> bool {
    matches!(
        expr,
        Expr::Member(MemberExpr {
            obj:
                box Expr::MetaProp(MetaPropExpr {
                    kind: MetaPropKind::ImportMeta,
                    ..
                }),
            prop:
                MemberProp::Ident(Ident {
                    sym: js_word!("url"),
                    ..
                }),
            ..
        })
    )
}

pub fn is_dynamic_import(call_expr: &CallExpr) -> bool {
    matches!(&call_expr.callee, Callee::Import(Import { .. }))
}

pub fn is_commonjs_require(call_expr: &CallExpr, unresolved_mark: &Mark) -> bool {
    if let Callee::Expr(box Expr::Ident(ident)) = &call_expr.callee {
        ident.sym == *"require" && is_native_ident(ident, unresolved_mark)
            || ident.sym == *"__mako_require__"
    } else {
        false
    }
}

pub fn is_native_ident(ident: &Ident, unresolved_mark: &Mark) -> bool {
    let outer = ident.span.ctxt.outer();

    outer == *unresolved_mark
}

pub fn get_first_arg_str(call_expr: &CallExpr) -> Option<String> {
    if let Some(arg) = call_expr.args.first() {
        if let box Expr::Lit(Lit::Str(str_)) = &arg.expr {
            return Some(str_.value.to_string());
        }
    }
    None
}
