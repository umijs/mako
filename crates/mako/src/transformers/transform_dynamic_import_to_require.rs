use mako_core::swc_common::{Mark, DUMMY_SP};
use mako_core::swc_ecma_ast::{Expr, ExprOrSpread, Lit};
use mako_core::swc_ecma_utils::{member_expr, quote_ident, quote_str, ExprFactory};
use mako_core::swc_ecma_visit::{VisitMut, VisitMutWith};

use crate::plugins::javascript::is_dynamic_import;

pub struct DynamicImportToRequire {
    pub unresolved_mark: Mark,
}

// import('xxx') -> Promise.resolve().then(() => require('xxx'))
impl VisitMut for DynamicImportToRequire {
    fn visit_mut_expr(&mut self, expr: &mut Expr) {
        if let Expr::Call(call_expr) = expr {
            if is_dynamic_import(call_expr) {
                if let ExprOrSpread {
                    expr: box Expr::Lit(Lit::Str(ref mut source)),
                    ..
                } = &mut call_expr.args[0]
                {
                    // Promise.resolve()
                    let promise_resolve: Box<Expr> = member_expr!(DUMMY_SP, Promise.resolve)
                        .as_call(DUMMY_SP, vec![])
                        .into();

                    // () => require( source.value... )
                    let lazy_require =
                        quote_ident!(DUMMY_SP.apply_mark(self.unresolved_mark), "require")
                            .as_call(DUMMY_SP, vec![quote_str!(source.value.clone()).as_arg()])
                            .into_lazy_arrow(vec![]);

                    // Promise.resolve().then(() => require("xxx"))
                    let promised_lazy_require: Expr =
                        member_expr!(@EXT,DUMMY_SP, promise_resolve, then)
                            .as_call(DUMMY_SP, vec![lazy_require.as_arg()]);

                    *expr = promised_lazy_require;
                }
            }
        }
        expr.visit_mut_children_with(self);
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use mako_core::swc_common::{Mark, GLOBALS};

    use crate::compiler::Context;

    #[test]
    fn test_basic_import() {
        crate::assert_display_snapshot!(transform(
            r#"
const testModule = import('test-module');
            "#,
        ));
    }

    #[test]
    fn test_chained_import() {
        crate::assert_display_snapshot!(transform(
            r#"
import('test-module').then(() => (
import('test-module-2')
));

Promise.all([
import('test-1'),
import('test-2'),
import('test-3'),
]).then(() => {});
            "#,
        ));
    }

    // TODO: support this?
    #[test]
    fn test_dynamic_import() {
        crate::assert_display_snapshot!(transform(
            r#"
import(MODULE);

let i = 0;
import(i++);

import(fn());

async () => import(await "x");

function* f() { import(yield "x"); }
            "#,
        ));
    }

    #[test]
    fn test_import_with_comment() {
        crate::assert_display_snapshot!(transform(
            r#"
import(/* test comment */ 'my-module');
import('my-module' /* test comment */ );
            "#,
        ));
    }

    // TODO: support this?
    #[test]
    fn test_template_argument() {
        crate::assert_display_snapshot!(transform(
            r#"
import(`1`);
import(tag`2`);
import(`3-${MODULE}`);
            "#,
        ));
    }

    fn transform(code: &str) -> String {
        let context: Arc<Context> = Arc::new(Default::default());
        GLOBALS.set(&context.meta.script.globals, || {
            let mut visitor = super::DynamicImportToRequire {
                unresolved_mark: Mark::new(),
            };
            crate::transformers::test_helper::transform_js_code(code, &mut visitor, &context)
        })
    }
}
