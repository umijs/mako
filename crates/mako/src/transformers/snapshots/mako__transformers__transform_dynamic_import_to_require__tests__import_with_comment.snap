---
source: crates/mako/src/transformers/transform_dynamic_import_to_require.rs
expression: "transform(r#\"\nimport(/* test comment */ 'my-module');\nimport('my-module' /* test comment */ );\n            \"#)"
---
Promise.resolve().then(()=>require("my-module"));
Promise.resolve().then(()=>require("my-module"));
